File: 2021/08/18/BinaryTreePaths.java
/**
 * https://leetcode.com/problems/binary-tree-paths/
 * @author Andrew Jarombek
 * @since 8/18/2021
 */

// Ran the 1500m on the Outdoor Track.  4:32, so nothing fast but it was a lot of fun.

import java.util.ArrayList;
import java.util.List;

public class BinaryTreePaths {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {}

        TreeNode(int val) { this.val = val; }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<String> binaryTreePaths(TreeNode root) {
        List<String> paths = new ArrayList<>();
        traverseNode(root, String.valueOf(root.val), paths);
        return paths;
    }

    public void traverseNode(TreeNode node, String path, List<String> paths) {
        if (node.left == null && node.right == null) {
            paths.add(path);
        } else {
            if (node.left != null) {
                traverseNode(node.left, path + "->" + String.valueOf(node.left.val), paths);
            }

            if (node.right != null) {
                traverseNode(node.right, path + "->" + String.valueOf(node.right.val), paths);
            }
        }
    }
}

File: 2022/02/24/binaryTreePaths.ts
/**
 * https://leetcode.com/problems/binary-tree-paths/
 * @author Andrew Jarombek
 * @since 2/24/2022
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

function binaryTreePaths(root: TreeNode | null): string[] {
    if (!root) {
        return [];
    }

    if (!root.left && !root.right) {
        return [`${root.val}`];
    }

    return binaryTreePaths(root.left)
        .concat(binaryTreePaths(root.right))
        .map(path => `${root.val}->${path}`);
}